pkg/controller/daemon/update.go

```diff
@@ -0,0 +1,138 @@
+
+	maxUnavailable, numUnavailable, err := dsc.getUnavailableNumbers(ds, allPods)
+	if err != nil {
+		glog.Errorf("Couldn't get unavailable numbers: %v", err)
+		return nil
+	}
+	oldAvailablePods, oldUnavailablePods := util.SplitByAvailablePods(ds.Spec.MinReadySeconds, oldPods)
+
+	// for oldPods delete all not running pods
+	var podsToDelete []string
+	glog.V(4).Infof("Marking all unavailable old pods for deletion")
+	for _, pod := range oldUnavailablePods {
+		// Skip terminating pods. We won't delete them again
+		if pod.DeletionTimestamp != nil {
+			continue
+		}
+		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
+		podsToDelete = append(podsToDelete, pod.Name)
+	}
+
+	glog.V(4).Infof("Marking old pods for deletion")
+	for _, pod := range oldAvailablePods {
+		if numUnavailable >= maxUnavailable {
+			glog.V(4).Infof("Number of unavailable DaemonSet pods: %d, is equal to or exceeds allowed maximum: %d", numUnavailable, maxUnavailable)
+			break
+		}
+		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
+		podsToDelete = append(podsToDelete, pod.Name)
+		numUnavailable++
+	}
+	errors := dsc.syncNodes(ds, podsToDelete, []string{})
+	return utilerrors.NewAggregate(errors)
+}
+
+func (dsc *DaemonSetsController) getAllDaemonSetPods(ds *extensions.DaemonSet) ([]*v1.Pod, []*v1.Pod, error) {
+	var newPods []*v1.Pod
+	var oldPods []*v1.Pod
+
+	selector, err := metav1.LabelSelectorAsSelector(ds.Spec.Selector)
+	if err != nil {
+		return newPods, oldPods, err
+	}
+	daemonPods, err := dsc.podLister.Pods(ds.Namespace).List(selector)
+	if err != nil {
+		return newPods, oldPods, fmt.Errorf("Couldn't get list of pods for daemon set %#v: %v", ds, err)
+	}
+	for _, pod := range daemonPods {
+		if util.IsPodUpdated(ds.TemplateGeneration, pod) {
+			newPods = append(newPods, pod)
+		} else {
+			oldPods = append(oldPods, pod)
+		}
+	}
+	return newPods, oldPods, nil
+}
+
+func (dsc *DaemonSetsController) getUnavailableNumbers(ds *extensions.DaemonSet, allPods []*v1.Pod) (int, int, error) {
+	glog.V(4).Infof("Getting unavailable numbers")
+	// TODO: get nodeList once in syncDaemonSet and pass it to other functions
+	nodeList, err := dsc.nodeLister.List(labels.Everything())
+	if err != nil {
+		return -1, -1, fmt.Errorf("couldn't get list of nodes during rolling update of daemon set %#v: %v", ds, err)
+	}
+
+	nodeToDaemonPods, err := dsc.getNodesToDaemonPods(ds)
+	if err != nil {
+		return -1, -1, fmt.Errorf("error getting node to daemon pod mapping for daemon set %#v: %v", ds, err)
+	}
+
+	var numUnavailable, desiredNumberScheduled int
+	for i := range nodeList {
+		node := nodeList[i]
+		wantToRun, _, _, err := dsc.nodeShouldRunDaemonPod(node, ds)
+		if err != nil {
+			return -1, -1, err
+		}
+		if wantToRun {
+			desiredNumberScheduled++
+			daemonPods, exists := nodeToDaemonPods[node.Name]
+			if !exists {
+				numUnavailable++
+				continue
+			}
+			available := false
+			for _, pod := range daemonPods {
```

Comment:
We shouldn't have mutliple pods here - make sure you don't count something as available and then delete it as a duplicate.